enable non-standard tilesfile

sbailey opened the pull request:
This PR fixes #63 to enable loading non-standard tiles files via the `desimodel.io.load_tiles` interface, albeit with a slightly different API:
```python
def load_tiles(onlydesi=True, extra=False, tilesfile=None, cache=True):
    """Return DESI tiles structure from desimodel/data/footprint/desi-tiles.fits.

    Parameters
    ----------
    onlydesi : :class:`bool` (default True)
        If ``True``, trim to just the tiles in the DESI footprint.
    extra : :class:`bool`, (default False)
        If ``True``, include extra layers with PROGRAM='EXTRA'.
    tilesfile : (str) name of tiles file to load; or None for default
        Without path, look in $DESIMODEL/data/footprint, otherwise load file
    cache : :class:`bool`, (default True)
        Use cache of tiles data
    """
```
i.e. 3 cases for `tilesfiles`:
* None: use default tiles file; handy for client code that wants to have an command line option for overriding the default tiles file without wanting to hard code that the current default in desimodel is called "desi-tiles.fits"
* filename including an absolute or relative path: load that file
* bare filename not including a path: look for a file of that name in $DESIMODEL/data/footprint.

Note: Use './desi-tiles.fits' to load a file in the pwd instead of $DESIMODEL/data/footprint/desi-tiles.fits .

Implementation note: the `_tiles` cache is now a dictionary keyed by the absolute path of the files that were loaded, enabling caching more than one file and distinguishing whether the requested file is in the cache or not.  This change resulted in this PR touching more lines of code than I had originally expected.

sbailey commented:
After some futzing around with tests and API tweaks, this is now ready for final review.

dkirkby commented on py/desimodel/io.py:
```diff
@@ -131,37 +131,57 @@ def load_tiles(onlydesi=True, extra=False):
+        tilesfile = 'desi-tiles.fits'
+
+    tilespath, filename = os.path.split(tilesfile)
+    if tilespath == '':
```
use of `os.path.isabs` would be more explicit here ([link](https://docs.python.org/3.5/library/os.path.html#os.path.isabs))

dkirkby commented on py/desimodel/io.py:
```diff
@@ -131,37 +131,57 @@ def load_tiles(onlydesi=True, extra=False):
         If ``True``, trim to just the tiles in the DESI footprint.
     extra : :class:`bool`, (default False)
         If ``True``, include extra layers with PROGRAM='EXTRA'.
+    tilesfile : (str) name of tiles file to load; or None for default
```
This should split into two lines if you are using numpy docstring conventions

dkirkby commented on py/desimodel/test/test_io.py:
```diff
@@ -2,7 +2,7 @@
 """Test desimodel.io.
 """
-import os
+import os, sys
```
https://www.python.org/dev/peps/pep-0008/#imports

dkirkby commented on py/desimodel/test/test_io.py:
```diff
@@ -2,7 +2,7 @@
 """
-import os
+import os, sys
 import uuid
```
I recommend using [tempfile](https://docs.python.org/3/library/tempfile.html) instead of rolling your own with `uuid`.  For example, see this [desisurvey unit test](https://github.com/desihub/desisurvey/blob/master/py/desisurvey/test/test_config.py).

sbailey commented:
I implemented the suggestions except for the suggestion to use `os.path.isabs` instead of `os.path.split()[0]`.  In this case I want to check if the input tilesfile has *any* path included, relative or absolute, not just absolute.  i.e. no path = look in $DESIMODEL/data/footprint/; with path (relative or absolute) = load the file as specified.
